lib/src/wallet/traits.rs

```diff
@@ -0,0 +1,590 @@
+    ShieldedOutput<D, ENC_CIPHERTEXT_SIZE>
+{
+    fn domain(&self, height: BlockHeight, parameters: P) -> D;
+    /// A decryption key for `enc_ciphertext`.  `out_ciphertext` is _itself_  decryptable
+    /// with the `OutgoingCipherKey` "`ock`".
+    fn out_ciphertext(&self) -> [u8; 80];
+    fn value_commitment(&self) -> D::ValueCommitment;
+}
+
+impl<A, P: Parameters> ShieldedOutputExt<P, OrchardDomain> for Action<A> {
+    fn domain(&self, _block_height: BlockHeight, _parameters: P) -> OrchardDomain {
+        OrchardDomain::for_action(self)
+    }
+
+    fn out_ciphertext(&self) -> [u8; 80] {
+        self.encrypted_note().out_ciphertext
+    }
+
+    fn value_commitment(&self) -> orchard::value::ValueCommitment {
+        self.cv_net().clone()
+    }
+}
+
+impl<P: Parameters> ShieldedOutputExt<P, SaplingDomain<P>> for OutputDescription<GrothProofBytes> {
+    fn domain(&self, height: BlockHeight, parameters: P) -> SaplingDomain<P> {
+        SaplingDomain::for_height(parameters, height)
+    }
+
+    fn out_ciphertext(&self) -> [u8; 80] {
+        self.out_ciphertext
+    }
+
+    fn value_commitment(&self) -> <SaplingDomain<Network> as Domain>::ValueCommitment {
+        self.cv
+    }
+}
+
+pub(crate) trait FromBytes<const N: usize> {
+    fn from_bytes(bytes: [u8; N]) -> Self;
+}
+
+impl FromBytes<32> for SaplingNullifier {
+    fn from_bytes(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl FromBytes<32> for OrchardNullifier {
+    fn from_bytes(bytes: [u8; 32]) -> Self {
+        Option::from(OrchardNullifier::from_bytes(&bytes))
+            .expect(&format!("Invalid nullifier {:?}", bytes))
+    }
+}
+
+pub(crate) trait FromCommitment {
+    fn from_commitment(from: &[u8; 32]) -> Self;
+}
+
+impl FromCommitment for SaplingNode {
+    fn from_commitment(from: &[u8; 32]) -> Self {
+        Self::new(*from)
+    }
+}
+impl FromCommitment for MerkleHashOrchard {
+    fn from_commitment(from: &[u8; 32]) -> Self {
+        Self::from_bytes(from).unwrap()
+    }
+}
+
+pub(crate) trait Spend {
+    type Nullifier: PartialEq + FromBytes<32> + UnspentFromWalletTxns;
+    fn nullifier(&self) -> &Self::Nullifier;
+    fn wallet_nullifier(_: &Self::Nullifier) -> WalletNullifier;
+}
+
+impl<Auth: SaplingAuthorization> Spend for SpendDescription<Auth> {
+    type Nullifier = SaplingNullifier;
+    fn nullifier(&self) -> &Self::Nullifier {
+        &self.nullifier
+    }
+    fn wallet_nullifier(null: &Self::Nullifier) -> WalletNullifier {
+        WalletNullifier::Sapling(*null)
+    }
+}
+
+impl<Auth> Spend for Action<Auth> {
+    type Nullifier = OrchardNullifier;
+    fn nullifier(&self) -> &Self::Nullifier {
+        self.nullifier()
+    }
+    fn wallet_nullifier(null: &Self::Nullifier) -> WalletNullifier {
+        WalletNullifier::Orchard(*null)
+    }
+}
+
+pub(crate) trait Recipient {
+    type Diversifier;
+    fn diversifier(&self) -> Self::Diversifier;
+    fn encode(&self, chain: Network) -> String;
```

Comment:
How constrained are we by this name?    I think a new type wrapper might be ideal for the returned value, and the method name might indicate how/why we provide this representation?

Reply:
We call functions that return strings, and usually we pass them to a user with no further transformation, where the only reason to turn them into strings is to write them down somewhere.. I think `String` is a fine type, and we've improved the name.